### ISSUE TITLE
ES5 "org.elasticsearch.index.mapper.MapperParsingException: failed to parse [fields.0]" on receiving end when using anonymous destructuring.

### ISSUE THREAD

**Issue by Rzpeg** [serilog/serilog-sinks-elasticsearch#76]
Hello,

Given the empty ES installation and the following console app code:

```
class Program
    {
        static void Main(string[] args)
        {
            Log.Logger = new LoggerConfiguration()
                .WriteTo.Elasticsearch(new ElasticsearchSinkOptions(new Uri("http://localhost:9200"))
                {
                    AutoRegisterTemplate = true
                })
                .CreateLogger();

            while (true)
            {
                Log.Information("Here is some information {@0}", new { NowIs = DateTime.UtcNow, SomeStuff = "something" });
                Thread.Sleep(300);
            }
        }
    }
```
The logs don't enter Elasticsearch 5 with the following exception:

```
org.elasticsearch.index.mapper.MapperParsingException: failed to parse [fields.0]
	at org.elasticsearch.index.mapper.FieldMapper.parse(FieldMapper.java:297) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:438) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:475) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:371) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:361) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrField(DocumentParser.java:435) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.DocumentParser.parseObject(DocumentParser.java:475) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.DocumentParser.innerParseObject(DocumentParser.java:371) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.DocumentParser.parseObjectOrNested(DocumentParser.java:361) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.DocumentParser.internalParseDocument(DocumentParser.java:93) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.DocumentParser.parseDocument(DocumentParser.java:66) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.DocumentMapper.parse(DocumentMapper.java:274) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.shard.IndexShard.prepareIndex(IndexShard.java:529) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.shard.IndexShard.prepareIndexOnPrimary(IndexShard.java:506) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.index.TransportIndexAction.prepareIndexOperationOnPrimary(TransportIndexAction.java:174) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.index.TransportIndexAction.executeIndexRequestOnPrimary(TransportIndexAction.java:179) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.bulk.TransportShardBulkAction.shardIndexOperation(TransportShardBulkAction.java:351) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.bulk.TransportShardBulkAction.index(TransportShardBulkAction.java:158) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.bulk.TransportShardBulkAction.handleItem(TransportShardBulkAction.java:137) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.bulk.TransportShardBulkAction.onPrimaryShard(TransportShardBulkAction.java:123) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.bulk.TransportShardBulkAction.onPrimaryShard(TransportShardBulkAction.java:74) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.TransportWriteAction.shardOperationOnPrimary(TransportWriteAction.java:78) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.TransportWriteAction.shardOperationOnPrimary(TransportWriteAction.java:50) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryShardReference.perform(TransportReplicationAction.java:902) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryShardReference.perform(TransportReplicationAction.java:872) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:113) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.onResponse(TransportReplicationAction.java:319) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.onResponse(TransportReplicationAction.java:254) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$1.onResponse(TransportReplicationAction.java:838) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$1.onResponse(TransportReplicationAction.java:835) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.shard.IndexShardOperationsLock.acquire(IndexShardOperationsLock.java:142) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.shard.IndexShard.acquirePrimaryOperationLock(IndexShard.java:1651) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.TransportReplicationAction.acquirePrimaryShardReference(TransportReplicationAction.java:847) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.doRun(TransportReplicationAction.java:271) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:250) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryOperationTransportHandler.messageReceived(TransportReplicationAction.java:242) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:69) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.transport.TransportService$6.doRun(TransportService.java:548) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:504) [elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37) [elasticsearch-5.0.0.jar:5.0.0]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:1.8.0_111]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:1.8.0_111]
	at java.lang.Thread.run(Unknown Source) [?:1.8.0_111]
Caused by: java.lang.IllegalStateException: Can't get text on a START_OBJECT at 1:231
	at org.elasticsearch.common.xcontent.json.JsonXContentParser.text(JsonXContentParser.java:85) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.common.xcontent.support.AbstractXContentParser.textOrNull(AbstractXContentParser.java:199) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.TextFieldMapper.parseCreateField(TextFieldMapper.java:379) ~[elasticsearch-5.0.0.jar:5.0.0]
	at org.elasticsearch.index.mapper.FieldMapper.parse(FieldMapper.java:286) ~[elasticsearch-5.0.0.jar:5.0.0]
	... 43 more
```

Changing "{@0}" to "{@Info}"  makes it work, however this is not the solution I am looking for, because of existing codebase, and other active sinks.

Any help appreciated.

P.S.
There are a lot of warnings for
`The [string] field is deprecated, please use [text] or [keyword] instead on *`

Regards,

**Comment by Rzpeg** [comment]
Forgot to add, I have the following packages installed

```
<packages>
  <package id="Elasticsearch.Net" version="2.4.6" targetFramework="net461" />
  <package id="Serilog" version="2.3.0" targetFramework="net461" />
  <package id="Serilog.Sinks.Elasticsearch" version="4.1.1" targetFramework="net461" />
  <package id="Serilog.Sinks.File" version="3.1.1" targetFramework="net461" />
  <package id="Serilog.Sinks.PeriodicBatching" version="2.1.0" targetFramework="net461" />
  <package id="Serilog.Sinks.RollingFile" version="3.2.0" targetFramework="net461" />
</packages>
```

**Comment by Rzpeg** [comment]
I have been toying with Elasticsearch API (via Sense) and if I run the following request without registering any templates, I get the entry created.

```
POST _bulk
{"index":{"_index":"logstash-2016.11.17","_type":"logevent"}}
{"@timestamp":"2016-11-17T14:23:57.9886572+00:00","level":"Information","messageTemplate":"Here is some stuff {@0}","message":"Here is some stuff { NowIs: 11/17/2016 14:23:57, SomeStuff: \"something\" }","fields":{"0":{"NowIs":"2016-11-17T14:23:57.9561564Z","SomeStuff":"something"}}}
```

```
{
   "took": 16,
   "errors": false,
   "items": [
      {
         "index": {
            "_index": "logstash-2016.11.17",
            "_type": "logevent",
            "_id": "AVhy0hLmTiBab5jV7DWH",
            "_version": 1,
            "result": "created",
            "_shards": {
               "total": 2,
               "successful": 1,
               "failed": 0
            },
            "created": true,
            "status": 201
         }
      }
   ]
}
```

If I run the sink's template, the same _bulk query results in error:

```
{
   "took": 331,
   "errors": true,
   "items": [
      {
         "index": {
            "_index": "logstash-2016.11.17",
            "_type": "logevent",
            "_id": "AVhy3diRTiBab5jV7DWQ",
            "status": 400,
            "error": {
               "type": "mapper_parsing_exception",
               "reason": "failed to parse [fields.0]",
               "caused_by": {
                  "type": "illegal_state_exception",
                  "reason": "Can't get text on a START_OBJECT at 1:219"
               }
            }
         }
      }
   ]
}
```

The created template is :

```
{
  "serilog-events-template": {
    "order": 0,
    "template": "logstash-*",
    "settings": {
      "index": {
        "refresh_interval": "5s"
      }
    },
    "mappings": {
      "_default_": {
        "dynamic_templates": [
          {
            "numerics_in_fields": {
              "match_pattern": "regex",
              "path_match": "fields\\.[\\d+]$",
              "mapping": {
                "index": "analyzed",
                "omit_norms": true,
                "type": "string"
              }
            }
          },
          {
            "string_fields": {
              "mapping": {
                "index": "analyzed",
                "omit_norms": true,
                "type": "string",
                "fields": {
                  "raw": {
                    "ignore_above": 256,
                    "index": "not_analyzed",
                    "type": "string"
                  }
                }
              },
              "match_mapping_type": "string",
              "match": "*"
            }
          }
        ],
        "_all": {
          "omit_norms": true,
          "enabled": true
        },
        "properties": {
          "message": {
            "index": "analyzed",
            "type": "string"
          },
          "exceptions": {
            "type": "nested",
            "properties": {
              "ExceptionMessage": {
                "type": "object",
                "properties": {
                  "MemberType": {
                    "type": "integer"
                  }
                }
              },
              "StackTraceString": {
                "index": "analyzed",
                "type": "string"
              },
              "HResult": {
                "type": "integer"
              },
              "RemoteStackTraceString": {
                "index": "analyzed",
                "type": "string"
              },
              "RemoteStackIndex": {
                "type": "integer"
              },
              "Depth": {
                "type": "integer"
              }
            }
          }
        }
      }
    },
    "aliases": {}
  }
}
```

**Comment by mivano** [comment]
The default template mapping tries to force the default 0, 1, etc fields to be strings in order to support old applications where you log for example:

```csharp
 Log.Information("log line 1", 29292);
 Log.Information("log line 2", "text");
```

Without the mapping, the fields.0 will become a integer field (as that is the first event that gets consumed) and then ES will fail to parse the string. So we force the 0,1,2, etc to be a string. You try however to fit an object in there which cannot be parsed.

So indeed use another name, add a destructor for these types or change the default template.

**Comment by mivano** [comment]
The ES mapper warnings are fixed in v5.7
